### PR TITLE
[SYSTEMDS-3553] Additional DNN/factorization optimizers and preconditioners (AdamW, ScaledGD)

### PR DESCRIPTION
**Pull request by ReneEnjilian**
This patch implements two new optimization algorithms: 

- Adam with weight decay (AdamW): https://arxiv.org/abs/1711.05101v3
- Scaled Gradient Descent (ScaledGD): https://arxiv.org/pdf/2005.08898

I did not add the shampoo optimizer because another student is working on that.

### CODE REVIEW

**Comment by mboehm7** [comment]
LGTM - thanks for the patch @ReneEnjilian. During the merge, I fixed the formatting and it would be nice to add tests in a future PR and report on the convergence speed of these optimizers.
